### ISSUE TITLE
Loss function of the CT algo variant

### ISSUE THREAD

**Issue by dimarkov** [bayesianempirimancer/jax-noprop#4]
If we look at the variational free energy for the CT variant it is expressed as 
$$E_{q(z_1|y)}\left[ - \ln p(y|z_1)\right] + D_KL\left( q(z_0|y)||p(z_0) \right) + \frac{\eta}{2}E_{t \sim \mathcal{U}(0, 1)} \left[ SNR^\prime(t) || u_\theta(z_t, x, t)  - u_y||^2\right] $$

So the overall objective can be split into three components: 
 *  $E_{q(z_1|y)}\left[ - \ln p(y|z_1)\right]  \approx - \ln p(y|\sqrt{\bar{\alpha}(1)} u_y + \sqrt{1 - \bar{\alpha}(1)} \epsilon), \quad \epsilon \sim \mathcal{N}(0, 1) $. Note that in general $\alpha(1) < 1$ unless explicitly specified with the noise schedule.
* $D_{KL}\left( q(z_0|y)||p(z_0) \right) = \frac{\bar{\alpha}(0)}{2} ||u_y||^2  -  \frac{D}{2} \left( \bar{\alpha}(0) + \ln (1 - \bar{\alpha}(0)) \right)$, where $D$ is the dimensionality of the encoding $u_y$. This terms sets the desired norm of the encoding $u_y$ if $\bar{alpha}(0) > 0$.
* the flow loss term which is not the issue here. 

Given this decomposition the loss should be computed differently then what is implemented in
```python
@partial(jax.jit, static_argnums=(0, 5))
    def loss(self, params: dict, x: Optional[jnp.ndarray], y: jnp.ndarray, key: jr.PRNGKey, training: bool = True) -> Tuple[jnp.ndarray, dict]:
       ...
        # Split keys for random sampling operations (not dropout)
        key, t_key, noise_key, z_target_key, vae_noise_key = jr.split(key, 5)
        batch_shape = y.shape[:-self.y_ndims]
        alpha_1 = self.apply(params, jnp.array(1.0), method='get_noise_params')[0]

        # Encode Target (noisy latent)
        mu_z_target, logvar_z_target = self.apply(params, y, method='encode', training=training, rngs={'dropout': key})
        z_target = mu_z_target + jr.normal(z_target_key, mu_z_target.shape) * jnp.exp(0.5 * logvar_z_target)
        
        # Sample time and get noise schedule parameters
        t = jr.uniform(t_key, batch_shape, minval=0.0, maxval=1.0)
        t_expanded = jnp.expand_dims(t, axis=tuple(range(-self.z_ndims, 0)))       
        # Get noise schedule parameters (expand t to match expected shape)
        alpha_t, gamma_prime_t = self.apply(params, t_expanded, method='get_noise_params')
        
        sqrt_alpha_t = jnp.sqrt(alpha_t)
        sqrt_1_minus_alpha_t = jnp.sqrt(1.0 - alpha_t)
        
        # Compute noisy latent state at time t
        z_t = sqrt_alpha_t * z_target + sqrt_1_minus_alpha_t * jr.normal(noise_key, z_target.shape)
        
        # Compute Target estimate and flow prediction
        z_target_est = self.apply(params, z_t, x, t, method='crn_output', training=training, rngs={'dropout': key})
        dz_dt = self.lazy_flow(z_t, z_target_est, alpha_t, gamma_prime_t)

        # Compute Predictions
        y_pred = self.apply(params, z_target_est, method='decode', training=training, rngs={'dropout': key})
        z_target_vae = z_target*jnp.sqrt(alpha_1) + jr.normal(vae_noise_key, z_target.shape) * jnp.sqrt(1.0 - alpha_1)
        y_vae = self.apply(params, z_target_vae, method='decode', training=training, rngs={'dropout': key})
        
        # Squeeze alpha_t and gamma_prime_t for use in SNR weight computation
        alpha_t_squeezed = jnp.squeeze(alpha_t, axis=tuple(range(-self.z_ndims, 0)))
        gamma_prime_t_squeezed = jnp.squeeze(gamma_prime_t, axis=tuple(range(-self.z_ndims, 0)))
        
        # Compute SNR weight (always computed, used for weighting)
        snr_weight = self.lazy_target_snr(alpha_t_squeezed, gamma_prime_t_squeezed)
        # Clip SNR weight to prevent explosion when alpha_t is close to 1
        # Maximum reasonable SNR weight: clip to 100 to prevent numerical issues
        snr_weight = jnp.clip(snr_weight, 0.0, 100.0)
        # Normalize SNR weights by their mean if normalize_snr_weight is True
        if normalize_snr_weight:
            snr_weight_mean = jnp.mean(snr_weight)
            snr_weight = snr_weight / (snr_weight_mean + 1e-8)

        # Compute Losses
        squared_error = jnp.mean((z_target_est - z_target) ** 2, axis=tuple(range(-self.z_ndims, 0)))
        flow_loss = jnp.mean(snr_weight * squared_error)
        reg_loss = jnp.mean(dz_dt**2)

        if recon_loss_type == "cross_entropy":
            recon_loss = jnp.sum(-y * jnp.log(y_pred + 1e-8), axis = tuple(range(-self.y_ndims, 0)))
            vae_loss   = jnp.sum(-y * jnp.log(y_vae + 1e-8), axis = tuple(range(-self.y_ndims, 0)))
        elif recon_loss_type == "mse":
            recon_loss = jnp.sum((y - y_pred)**2, axis=tuple(range(-self.y_ndims, 0)))
            vae_loss   = jnp.sum((y - y_vae)**2, axis=tuple(range(-self.y_ndims, 0)))
        else:
            recon_loss = 0.0
            vae_loss = 0.0
        
        recon_loss = jnp.mean(snr_weight * recon_loss) 
        vae_loss = jnp.mean(vae_loss)

        total_loss = flow_loss + recon_weight * recon_loss + reg_weight * reg_loss + vae_weight * vae_loss
        
        return total_loss, {
            'flow_loss': flow_loss,
            'recon_loss': recon_loss, 
            'reg_loss': reg_loss,
            'vae_loss': vae_loss,
            'total_loss': total_loss
        }
```

First recon loss seems redundant, as the correct loss term is vae_loss. Second, the KL div term that sets the scale of the embedding z_target, is missing. Also, I am not sure where reg_loss terms is coming from as it does not come out of derivations in the NoProp paper.

**Comment by bayesianempirimancer** [comment]
implemented kl q(z_0|z_T), p(z_0) loss contribution.  closing.
